Re-add flag to hide the extensions menu (the puzzle piece icon)

jwangac opened the pull request:
This PR replaces my previous PR #1994. The previous PR is reverted because it
causes ungoogled-chromium to crash.

This PR re-adds the flag that allows one to hide the extensions menu (the
puzzle piece icon), by patching [the menu's visibility logic](https://github.com/chromium/chromium/blob/0f4ff69f75ce13ac45815a53000c36e7401561a5/chrome/browser/ui/views/extensions/extensions_toolbar_container.cc#L804):

```patch
 bool ExtensionsToolbarContainer::ShouldContainerBeVisible() const {
+  if (base::CommandLine::ForCurrentProcess()->HasSwitch("hide-extensions-menu"))
+    return false;
+
   // The container (and extensions-menu button) should not be visible if we have
   // no extensions.
   if (!HasAnyExtensions())
     return false;

   // All other display modes are constantly visible.
   if (display_mode_ != DisplayMode::kAutoHide)
     return true;

   if (GetAnimatingLayoutManager()->is_animating())
     return true;

   // Is menu showing.
   if (GetExtensionsButton()->GetExtensionsMenuShowing())
     return true;

   // Is extension pop out is showing.
   if (popped_out_action_)
     return true;

   // Is extension pop up showing.
   if (popup_owner_)
     return true;

   return false;
 }
```

I test the patch on [Windows](https://github.com/jwangac/ungoogled-chromium-windows/releases/tag/103.0.5060.53-test). The patch works and avoids previous crashes

1. when installing a packed extension
2. when loading an unpacked extension
3. when packing an extension

(Thanks Rusenche for reporting these crash cases.)

But my test can be incomplete as I'm not a heavy extension user or developer.
This PR is marked as draft. I'm wondering if someone can help testing it in 
more use-cases.

This patch has some limitations.

1. The flag will hide all pinned extensions. One cannot hide only the puzzle
piece icon but keep pinned extensions.

2. Notification like `[Extension Name] has been added to Chromium` (shows after
installing an extension) and `Remove "[Extension Name]"?` (shows when removing
an extension) cannot find the hidden extension icon to attach. The notification
will show at the left-top corner.

PF4Public commented:
@jwangac Have you changed your mind?

jwangac commented:
I've tested this patch for a while and would like to re-open this PR for
consideration.

Since the previous close of this PR, I updated the patch to fix the location of
extension-installed-bubble (`Add [extension name]? It can: ...`) and
extension-uninstall-dialog (`Remove [extension name]?`).

```patch
@@ -318,6 +319,18 @@ void ExtensionsToolbarContainer::AnchorA
   widget->widget_delegate()->AsBubbleDialogDelegate()->SetAnchorView(
       anchor_view && anchor_view->GetVisible() ? anchor_view
                                                : GetExtensionsButton());
+
+  // Fix the position of widgets. Without this fix, extension-installed-bubble
+  // and extension-uninstall-dialog may be out of the window border on Linux.
+  if (base::CommandLine::ForCurrentProcess()->HasSwitch("hide-extensions-menu"))
+  {
+    BrowserView* browser_view = BrowserView::GetBrowserViewForBrowser(browser_);
+    views::View* reference_view = browser_view->toolbar_button_provider()
+      ->GetAnchorView(PageActionIconType::kBookmarkStar);
+    widget->widget_delegate()->AsBubbleDialogDelegate()
+      ->SetAnchorView(reference_view);
+  }
+
   widget->Show();
 }
```

It is because these dialogue windows are anchored to the extension menu button.
If the extension menu button is hidden, these dialogue windows will show at 
wrong position; and the location is OS-dependent.

On Windows, these windows show at left-top corner. I guess it is acceptable
and does not require the above patch.

On Linux (specially Debian+Sway), these windows show at left-top corner, but are
out of the border of chromium and thus merely visible.

So I introduce the above patch to anchor these windows to the bookmark button,
which generates a similar position as if the extensions menu is not hidden.

While this patch works fine on Windows and Linux, I'm not very confident about
its behavior on other OS (like MacOS), and would like to know if there is a
better/clever way to fix the window position.

This patch has been tested on Windows and Linux (Debian + Sway).
- [x] 104.0.5112.81 ([Windows](https://github.com/jwangac/ungoogled-chromium-windows/releases/tag/104.0.5112.81) and [Portable-Linux](https://github.com/jwangac/ungoogled-chromium-portablelinux/releases/tag/104.0.5112.81))
- [x] 104.0.5112.102 ([Windows](https://github.com/jwangac/ungoogled-chromium-windows/releases/tag/104.0.5112.102) and [Portable-Linux](https://github.com/jwangac/ungoogled-chromium-portablelinux/releases/tag/104.0.5112.102-1.1))
- [x] 105.0.5195.102 ([Windows](https://github.com/jwangac/ungoogled-chromium-windows/releases/tag/105.0.5195.102-1) and [Portable-Linux](https://github.com/jwangac/ungoogled-chromium-portablelinux/releases/tag/105.0.5195.102-1.1))

PS: On Debian, I experience rare freeze on a specific website: I can control
chromium with its hotkey and vimimum's hotkey, but the screen did not change
anymore. As the freeze also happens when this patch is not enabled, I guess the
freeze is not induced by this patch, but can not locate real cause of freeze.

Ahrotahn commented:
I checked this out and found no issues.  The only thing I would suggest is it would probably be a better idea to set the anchor to the menu button [like the crash bubble](https://source.chromium.org/chromium/chromium/src/+/refs/tags/105.0.5195.102:chrome/browser/ui/views/session_crashed_bubble_view.cc;l=259).  You would have to add [an extra include](https://source.chromium.org/chromium/chromium/src/+/refs/tags/105.0.5195.102:chrome/browser/ui/views/session_crashed_bubble_view.cc;l=31), but that should future-proof things better since it's possible for the star button to be hidden like the sensor/cookie/QR/etc buttons.

PF4Public commented:
In addition to what **Ahrotahn** pointed out, please rebase your changes and also replace double-spaces with a single one for a better uniformity .

jwangac commented:
I accidentally closed this PR when making rebase. Sorry for the inconvenence.

@Ahrotahn Following your suggestion, I have set the anchor to the menu button.
Thanks for this suggestion and the detailed hint.

@PF4Public I have rebased my changes, sorted the "docs/flags.md" file, and
removed double-spaces in the "chrome/browser/ungoogled_flag_entries.h" patch.
Please let me know if I miss something.

The new patch is developed and briefly tested with portable linux. Fully rebuild 
and test may be completed in the next day.

Ahrotahn reviewed (approved):
No worries, I've done the same thing before.  
I checked out the new changes and this looks good to me!

PF4Public commented:
@Ahrotahn In your opinion, could our checks `base::CommandLine::ForCurrentProcess()->HasSwitch()` and `base::FeatureList::IsEnabled()` somehow affect performance?

@jwangac have you attempted a full rebuild and test yet?

jwangac commented:
> @jwangac have you attempted a full rebuild and test yet?

Yes, the new patch works as intended on [Portable Linux](https://github.com/jwangac/ungoogled-chromium-portablelinux/releases/tag/105.0.5195.102-1.1-alpha) and [Windows](https://github.com/jwangac/ungoogled-chromium-windows/releases/tag/105.0.5195.102-1.1).

Ahrotahn commented:
@PF4Public They're pretty efficient and wouldn't make any noticeable impact.  What matters more is *where* we make changes.  You would want to minimize conditionals in code that gets run frequently.

PF4Public commented:
Thanks @jwangac! Merged.

Marek33 commented:
@Ahrotahn Thank you very much for this excellent patch, but after I enable it, instead of just removing the puzzle icon, the extension icons are gone as well. Is that intended, or is it a bug?

_I know that I'm a bit late, but I forgot about this, and I'm sorry, for not reporting it soon._

jwangac commented:
> @Ahrotahn Thank you very much for this excellent patch, but after I enable it, instead of just removing the puzzle icon, the extension icons are gone as well. Is that intended, or is it a bug?
> 
> _I know that I'm a bit late, but I forgot about this, and I'm sorry, for not reporting it soon._

It is intended. This patch removes the extension menu as a whole; both the puzzle icon and pinned extension icons are hidden. (I have not found a safe way to hide the puzzle icon alone.)

The comments may be misleading:
```patch
+  `--hide-extensions-menu` | Hides the extensions menu (the puzzle piece icon).
```
```patch
+    {"hide-extensions-menu",
+     "Hide Extensions Menu",
+     "Hides the extensions menu (the puzzle piece icon). ungoogled-chromium flag.",
+     kOsDesktop, SINGLE_VALUE_TYPE("hide-extensions-menu")},
```
Maybe the text "(the puzzle piece icon)" should be removed?

Ahrotahn commented:
Maybe something like this would work?
> Hides the extensions container.  This includes the puzzle piece icon as well as any pinned extensions.  

The recent fullscreen ui patch needs to have its documentation updated as well, so that might be something we could combine into a PR to update both.
